### ISSUE TITLE
slow job submission with awsbatch executor

### ISSUE THREAD

**Issue by tbugfinder** [nextflow-io/nextflow#709]
I would like to submit >20000 jobs (one per each file in a S3 bucket) for parallel processing.

In average nextflow submission to awsbatch only creates ~1 job per second to awsbatch which is too slow for this kind of large scale szenario.
Each files is processed within 5-30min. This means in a horizontal scaling that the overall processing time could be executed in ~30min (plus overhead for instance creation and reporting).

nextflow trace:

```
May-18 14:43:52.559 [Task submitter] TRACE n.executor.AwsBatchFileCopyStrategy - [AWS BATCH] Unstaging file path: [RESULT.gz]

May-18 14:43:52.718 [Task submitter] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] new job request > {JobName: myjobname,JobQueue: myjobqueue,JobDefinition: myjobdefinition,ContainerOverrides: {Command: [bash, -o, pipefail, -c, trap "{ ret=$?; aws s3 cp --only-show-errors .command.log s3://testbucket123/work/sum-97eccecc-d7e9-431a-a3bc-20440c5c77e7/b0/f3781d20e330ae13258b7c15157b64/.command.log||true; exit $ret; }" EXIT; aws s3 cp --only-show-errors s3://testbucket123/work/sum-97eccecc-d7e9-431a-a3bc-20440c5c77e7/b0/f3781d20e330ae13258b7c15157b64/.command.run - | bash 2>&1 | tee .command.log],},RetryStrategy: {Attempts: 4},Timeout: {AttemptDurationSeconds: 108000}}

May-18 14:43:52.748 [Task submitter] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] submitted > job=78fd5324-1fee-4503-be6f-676d6606f631; work-dir=s3://testbucket123/work/sum-97eccecc-d7e9-431a-a3bc-20440c5c77e7/b0/f3781d20e330ae13258b7c15157b64

May-18 14:43:52.748 [Task submitter] INFO  nextflow.Session - [b0/f3781d] Submitted process > myjobname
```

using later version:
```
  Version: 0.30.0-SNAPSHOT build 4813
  Modified: 18-05-2018 15:51 UTC
  System: Linux 4.9.85-38.58.amzn1.x86_64
  Runtime: Groovy 2.4.15 on OpenJDK 64-Bit Server VM 1.8.0_171-b10
  Encoding: UTF-8 (UTF-8)
  Process: 13655@ip-10-4-13-234 [10.4.13.234]
  CPUs: 8 - Mem: 14.7 GB (8.7 GB) - Swap: 0 (0)
```
.........

#to be added#

AWS Batch itself doesn't throttle the submission.
Test:
```
i="0"
time (while [ $i -lt 30000 ]
do
date
time  (aws batch submit-job \
                --region eu-west-1 \
                --job-name test_test_test \
                --job-queue spot40-queue \
                --job-definition myjobdefinition:1 \
                --retry-strategy attempts=3 \
                --parameters cmd=/bin/ls,batch_run_id=test > /dev/null )
let i=i+1
done
)
```
==> finished within 20 minutes.

**Comment by pditommaso** [comment]
Regarding this issue, it's important to take in consideration that for each task NF requires to make at least 4 api call: 
1. create the task work directory in the S3 bucket;
2. upload the task launcher in the S3 work dir; 
3. upload the task script in the S3 work dir; 
4. submit the job execution.  

Therefore it's surely going to be slower than a simple bash loop. 

However profiling the AWS Batch executor I've realised that there was a glitch causing the use of multipart upload API call which is quite expensive (~ 400ms for each file) 

![screen shot 2018-05-19 at 16 08 40](https://user-images.githubusercontent.com/816968/40279383-0dfa58ee-5c42-11e8-91eb-3500c081f784.png)

I've uploaded a new snapshot that uses a regular `putObject` instead of a multipart upload which takes half of the time to upload each file. 

You may want to give a try using refreshing the current version with the following command: 

```
CAPSULE_RESET=1 NXF_VER=0.30.0-SNAPSHOT nextflow info
```

then run NF as shown below: 

```
NXF_VER=0.30.0-SNAPSHOT nextflow run .. etc
```

I'm also wondering if you are running NF in a local machine or a Ec2 instance?

**Comment by pditommaso** [comment]
Requires #712

**Comment by tbugfinder** [comment]
I've updated nf to  Version: 0.30.0-SNAPSHOT build 4805 (??) -  Modified: 25-05-2018 20:40 UTC

First testing shows ~2.5 job submissions per second which is really better than before however not enabling large scale job submissons.

**Comment by pditommaso** [comment]
Is this running in the AWS network using a ec2 instance ?

**Comment by tbugfinder** [comment]
The slower jobs are fed into AWS batch, the slower AWS batch scales.

**Comment by tbugfinder** [comment]
The AWS Batch submission is running for several hours (starting Jun-02 18:58) by now. It slowed down tremendously.
[nf.tasks_in_the_submission_queue.1528122523.txt](https://github.com/nextflow-io/nextflow/files/2068655/nf.tasks_in_the_submission_queue.1528122523.txt)

**Comment by tbugfinder** [comment]
To me it looks like the type of calculation doesn't matter. I'm building a **test case** using just a `create file & sleep 1200` execution in combination with a `output_channel.collect().set{ all_files }`. 

As the verification of the test case takes some time to complete, I'll come back soon.

I noticed  several messages of type:
```
Jun-04 04:04:52.863 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=2f3c8f82-bdc6-4bfc-9401-8ea33cae88d0
Jun-04 04:04:52.864 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[2f3c8f82-bdc6-4bfc-9401-8ea33cae88d0]
Jun-04 04:04:52.892 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=2f3c8f82-bdc6-4bfc-9401-8ea33cae88d0
Jun-04 04:04:52.892 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[492f00fa-98b5-44f1-961b-b1fba5f1ba7b]
Jun-04 04:04:52.925 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=492f00fa-98b5-44f1-961b-b1fba5f1ba7b
Jun-04 04:04:52.925 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[492f00fa-98b5-44f1-961b-b1fba5f1ba7b]
Jun-04 04:04:52.948 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=492f00fa-98b5-44f1-961b-b1fba5f1ba7b
Jun-04 04:04:52.949 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[d0ab9f41-36bd-413d-a782-a6c8323f51f3]
Jun-04 04:04:52.971 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=d0ab9f41-36bd-413d-a782-a6c8323f51f3
Jun-04 04:04:52.971 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[d0ab9f41-36bd-413d-a782-a6c8323f51f3]
Jun-04 04:04:52.995 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=d0ab9f41-36bd-413d-a782-a6c8323f51f3
Jun-04 04:04:52.995 [Task monitor] TRACE n.executor.AwsBatchTaskHandler - [AWS BATCH] requesting describe jobs=[26c5b718-c226-4154-b155-b2f4adb615e3]
Jun-04 04:04:53.020 [Task monitor] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] cannot retrieve running status for job=26c5b718-c226-4154-b155-b2f4adb615e3
```

**Comment by pditommaso** [comment]
That's a bit weird, it tries to look up the job info but the service isn't returning such info

https://github.com/nextflow-io/nextflow/blob/a773a7fc032814fd10527701fdbbba0885d441c7/src/main/groovy/nextflow/executor/AwsBatchExecutor.groovy#L271-L276

**Comment by tbugfinder** [comment]
As it would not finish anyway and the my Yourkit couldn't connect to it, I've pressed `CTRL+C`.
After that NF was trying to terminate many jobs (`Jun-05 15:04:09.280 [main] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] killing job=8e782b5b-1211-45d3-bf39-fabc4629fe51`), although those jobs were already finished within AWS Batch. So it looks like NF lost state at some point.

Nevertheless, I restarted the processing again including Yourkit agent, and got again into the really slow mode.

```
![image](https://user-images.githubusercontent.com/15378237/41023118-c290677c-696b-11e8-87f3-2a542c05ea84.png)

```

**Comment by tbugfinder** [comment]
This setting is / was configured.
```
 submitRateLimit = '400 / 1 s'
```
Does it apply to submits only or to checks of running jobs? Will increase that one and run another test.

**Comment by pditommaso** [comment]
> After that NF was trying to terminate many jobs (Jun-05 15:04:09.280 [main] DEBUG n.executor.AwsBatchTaskHandler - [AWS BATCH] killing job=8e782b5b-1211-45d3-bf39-fabc4629fe51), although those jobs were already finished within AWS Batch 

This smell bad. 

Also could you share the memory snapshot produced by Yourkit, via dropbox or something similar?

**Comment by tbugfinder** [comment]
yes, I do  - it's filepath and filesize. Depending on the filesize I'm selecting a different job-definition, although this isn't a must have.

```
Channel
        .fromPath(params.s3datadir)
        .map { file -> tuple (file, file.size()) }
        .view()
        .set { inputChannel }
```

Meanwhile I changed instance type from c4 to  m4.2xlarge and increased Java memory settings.
```
export NXF_OPTS="-Xms1024m -Xmx8192m -agentpath:/opt/YourKit/bin/linux-x86-64/libyjpagent.so"
```

It looks like the workflow will complete however overall it is to slow (2-3 jobs/second in submission phase).


Even the build result process which is collecting all result files is running since hours.
```
process buildResult {
        publishDir '/efs/nextflowtest/output', mode: 'copy', overwrite: true
        publishDir 's3://s3backedteststore/workdone', mode: 'copy', overwrite: true

        input:
        file summarylist from all_data_list

        output:
        file "allresults_of_this_search.sdf"

        """
        echo "what did we get?"
        ls -l $summarylist
        zcat -c $summarylist >> allresults_of_this_search.sdf
        cp allresults_of_this_search.sdf /efs/nextflowtest/output/
        """
}
```

[nf.main.rocs.txt](https://github.com/nextflow-io/nextflow/files/2083067/nf.main.rocs.txt)

**Comment by tbugfinder** [comment]
Actually this is slightly off-topic however the buildResult step takes also quite long just collecting the files (I know I have it twice, right now) from S3 (~1file per second - filesize about 2k).
```
# still running
# ls -l /tmp/nxf.jUpQLUda4y/| wc -l 
80573

#S3 command
# /usr/bin/python /usr/bin/aws s3 cp --only-show-errors s3://nextflowtest/work/<snip>.sdf.gz RESULTPREFIX<snip>.sdf.gz
```

**Comment by tbugfinder** [comment]
Any chance to enable EFS (NFS) for the Batch Executor also ?

**Comment by pditommaso** [comment]
I've gave a look at your script. First the following is useless: 

```
output_channel = Channel.create()
output_channel.collect().set{ all_data_list }
```

You can just have 

```
        input:
        file summarylist from output_channel.collect()

```

However that `collect` causes all that paths to be retained in memory which could explain such memory pressure, considering the large number of input files you are using.

**Comment by pditommaso** [comment]
> Any chance to enable EFS (NFS) for the Batch Executor also ?

That should require a custom AMI, I think it should be possible but never investigated that.

**Comment by tbugfinder** [comment]
We do have a custom AMI however and job definition so that EFS is available however NF would not use this as it's depending on S3 workdir if the Batch executor is used (as per my understanding).

**Comment by tbugfinder** [comment]
Question to: 
```
input:
        file summarylist from output_channel.collect()`
```
Does it mean that a pre-channel definition creates and holds the list beforehand whereas an input definition doesn't ?

**Comment by pditommaso** [comment]
>  However NF would not use this as it's depending on S3 workdir if the Batch executor is used (as per my understanding).

Good point, you are right. It wont' work. 

> Does it mean that a pre-channel definition creates and holds the list beforehand whereas an input definition doesn't ?

No. Means that `collect` needs to return all entries altogether as a solo emission, therefore it needs to keep all of them in memory.

**Comment by tbugfinder** [comment]
>> However NF would not use this as it's depending on S3 workdir if the Batch executor is used (as per my understanding).

>Good point, you are right. It wont' work.

Could I file another RFE in that sense ?

**Comment by pditommaso** [comment]
Oops .. it looks this is caused by an empty file in the `bin` directory? is that possible ?

**Comment by pditommaso** [comment]
I've made another snapshot. Update your copy with this command: 

```
CAPSULE_RESET=1 NXF_VER=0.31.0-SNAPSHOT nextflow info
```

Then run it as before: 

```
NXF_VER=0.31.0-SNAPSHOT nextflow run .. etc
```

**Comment by tbugfinder** [comment]
Yes, removed three empty files (0 B) and started successfully.

**Comment by tbugfinder** [comment]
looks like some sort of multi-channel or similar is needed.

```

Caused by:
  Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: 26587acd-7e31-11e8-a9fb-a1809f51969e)

com.amazonaws.services.batch.model.AWSBatchException: Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: 26587acd-7e31-11e8-a9fb-a1809f51969e)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1632)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1304)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
        at com.amazonaws.services.batch.AWSBatchClient.doInvoke(AWSBatchClient.java:1268)
        at com.amazonaws.services.batch.AWSBatchClient.invoke(AWSBatchClient.java:1244)
        at com.amazonaws.services.batch.AWSBatchClient.executeSubmitJob(AWSBatchClient.java:1039)
        at com.amazonaws.services.batch.AWSBatchClient.submitJob(AWSBatchClient.java:1014)
        at nextflow.cloud.aws.batch.AwsBatchTaskHandler.submit(AwsBatchTaskHandler.groovy:270)
        at nextflow.processor.TaskPollingMonitor.submit(TaskPollingMonitor.groovy:210)
        at nextflow.processor.ParallelPollingMonitor.super$2$submit(ParallelPollingMonitor.groovy)
        at nextflow.processor.ParallelPollingMonitor$_submit_closure2.doCall(ParallelPollingMonitor.groovy:93)
        at nextflow.processor.ParallelPollingMonitor$_submit_closure2.call(ParallelPollingMonitor.groovy)
        at groovy.lang.Closure.run(Closure.java:499)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

```

**Comment by tbugfinder** [comment]
```
Jul-02 19:50:19.684 [pool-4-thread-15] DEBUG nextflow.Session - Session aborted -- Cause: Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID:
26587acd-7e31-11e8-a9fb-a1809f51969e)
Jul-02 19:50:19.712 [pool-4-thread-15] DEBUG nextflow.Session - The following nodes are still active:
[process] runROCS
  status=ACTIVE
  port 0: (queue) OPEN  ; channel: -
  port 1: (cntrl) -     ; channel: $

[process] buildResult
  status=ACTIVE
  port 0: (value) OPEN  ; channel: summarylist
  port 1: (cntrl) -     ; channel: $

Jul-02 19:50:19.715 [Task monitor] DEBUG n.processor.TaskPollingMonitor - <<< barrier arrives (monitor: awsbatch)
Jul-02 19:50:19.715 [main] DEBUG nextflow.Session - Session await > all process finished
Jul-02 19:50:19.715 [main] DEBUG nextflow.Session - Session await > all barriers passed
Jul-02 19:50:19.731 [Actor Thread 2] ERROR n.extension.DataflowExtensions - @unknown
com.amazonaws.SdkClientException: Failed to sanitize XML document destined for handler class com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser$ListBucketHandler
        at com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser.sanitizeXmlDocument(XmlResponsesSaxParser.java:219)
        at com.amazonaws.services.s3.model.transform.XmlResponsesSaxParser.parseListBucketObjectsResponse(XmlResponsesSaxParser.java:317)
        at com.amazonaws.services.s3.model.transform.Unmarshallers$ListObjectsUnmarshaller.unmarshall(Unmarshallers.java:70)
        at com.amazonaws.services.s3.model.transform.Unmarshallers$ListObjectsUnmarshaller.unmarshall(Unmarshallers.java:59)
        at com.amazonaws.services.s3.internal.S3XmlResponseHandler.handle(S3XmlResponseHandler.java:62)
        at com.amazonaws.services.s3.internal.S3XmlResponseHandler.handle(S3XmlResponseHandler.java:31)
        at com.amazonaws.http.response.AwsResponseHandlerAdapter.handle(AwsResponseHandlerAdapter.java:70)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleResponse(AmazonHttpClient.java:1547)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1272)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4365)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4312)
        at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4306)
        at com.amazonaws.services.s3.AmazonS3Client.listObjects(AmazonS3Client.java:840)
        at com.upplication.s3fs.AmazonS3Client.listObjects(AmazonS3Client.java:104)
        at com.upplication.s3fs.util.S3ObjectSummaryLookup.lookup(S3ObjectSummaryLookup.java:117)
        at com.upplication.s3fs.S3FileSystemProvider.readAttributes(S3FileSystemProvider.java:643)
        at java.nio.file.Files.readAttributes(Files.java:1737)
        at java.nio.file.Files.size(Files.java:2332)
        at org.codehaus.groovy.runtime.NioGroovyMethods.size(NioGroovyMethods.java:121)
        at sun.reflect.GeneratedMethodAccessor75.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.groovy.runtime.metaclass.ReflectionMetaMethod.invoke(ReflectionMetaMethod.java:54)
        at org.codehaus.groovy.runtime.metaclass.NewInstanceMetaMethod.invoke(NewInstanceMetaMethod.java:56)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1225)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1034)
        at groovy.runtime.metaclass.NextflowDelegatingMetaClass.invokeMethod(NextflowDelegatingMetaClass.java:79)
        at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:46)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
        at _nf_script_2fe60636$_run_closure1.doCall(_nf_script_2fe60636:54)

```

**Comment by pditommaso** [comment]
Well, now the problem is that AWS is not able to process at such throughput. You need to lower the `submitRateLimit`.

**Comment by tbugfinder** [comment]
Are there multiple tcp channels in use for S3 transfer, job-status and job-submission so that the limits are not summed up?
Using anything greater than `30 / 1s`  failed right after 150 jobs, `20 / 1s` and `15 / 1s` were running until about 1400/1200 job submissions  then it also was killed - `10 / 1s` were killed after 1600 job submissions.

**Comment by pditommaso** [comment]
This sounds very bad. No, job status is checked sequentially. Is the exception reported always the same? I mean `TooManyRequestsException`?

**Comment by tbugfinder** [comment]
Yes, always the same exception.

**Comment by tbugfinder** [comment]
However it really looks promising. 
We also run into that exception while submitting jobs using CLI and ended up with parallel aws-cli invocations.

**Comment by pditommaso** [comment]
Interesting, this is very important can you confirm that's the identical stack trace? That would mean the problem affects only the submit requests.

**Comment by tbugfinder** [comment]
--confirmed--

`25 / 1s`
```
Caused by:
  Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: 63dcaeb5-7e3a-11e8-a12d-93309b17d689)

com.amazonaws.services.batch.model.AWSBatchException: Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: 63dcaeb5-7e3a-11e8-a12d-93309b17d689
)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1632)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1304)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
```

`15 / 1s`
```

Caused by:
  Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: ee6cf0c3-7e3a-11e8-af29-657a01b8eb64)

com.amazonaws.services.batch.model.AWSBatchException: Too Many Requests (Service: AWSBatch; Status Code: 429; Error Code: TooManyRequestsException; Request ID: ee6cf0c3-7e3a-11e8-af29-657a01b8eb64
)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1632)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1304)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1058)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:743)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:717)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
        at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
        at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
        at com.amazonaws.services.batch.AWSBatchClient.doInvoke(AWSBatchClient.java:1268)
        at com.amazonaws.services.batch.AWSBatchClient.invoke(AWSBatchClient.java:1244)
        at com.amazonaws.services.batch.AWSBatchClient.executeSubmitJob(AWSBatchClient.java:1039)
        at com.amazonaws.services.batch.AWSBatchClient.submitJob(AWSBatchClient.java:1014)
        at nextflow.cloud.aws.batch.AwsBatchTaskHandler.submit(AwsBatchTaskHandler.groovy:270)
        at nextflow.processor.TaskPollingMonitor.submit(TaskPollingMonitor.groovy:210)

```

**Comment by tbugfinder** [comment]
Overall submission speed with `0.31.0-SNAPSHOT build 4926` went down to `5h` (pretty good). Although a single compute job, takes 5min - 60min the overall processing takes 11h26min.
Unfortunately I cannot use the html reports due to the number of jobs in order to dig into it (time consumption) deeper - any idea how to deal with that? However I see a lot of time spent in [Task Monitor].
How could I figure out if/how the overall time could be reduced in this "post processing"?

```
Jul-10 05:51:34.341 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 107602; name: runJOB; status: COMPLETED; exit: 0; error: -; workDir: s3://test/work/sum-3da14332-801e-4449-bca0-b7d576c85f4f/cb/ba3198856edfbc51cd5158a7ea942a]
Jul-10 05:51:34.341 [Task monitor] DEBUG nextflow.file.FileHelper - Path matcher not defined by 'S3FileSystem' file system -- using default default strategy
Jul-10 05:51:34.477 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 107604; name: runJOB; status: COMPLETED; exit: 0; error: -; workDir: s3://test/work/sum-3da14332-801e-4449-bca0-b7d576c85f4f/8f/ce59dfe151e5822a2db00362107654]
Jul-10 05:51:34.477 [Task monitor] DEBUG nextflow.file.FileHelper - Path matcher not defined by 'S3FileSystem' file system -- using default default strategy
Jul-10 05:51:34.602 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 107606; name: runJOB); status: COMPLETED; exit: 0; error: -; workDir: s3://test/work/sum-3da14332-801e-4449-bca0-b7d576c85f4f/fc/b10d8cda08085409c1bd93d6c00ad1]
Jul-10 05:51:34.602 [Task monitor] DEBUG nextflow.file.FileHelper - Path matcher not defined by 'S3FileSystem' file system -- using default default strategy
Jul-10 05:51:34.766 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[id: 107612; name: runJOB; status: COMPLETED; exit: 0; error: -; workDir: s3://test/work/sum-3da14332-801e-4449-bca0-b7d576c85f4f/35/2308dfcd85a0953e6cbe8972f4407f]
Jul-10 05:51:34.766 [Task monitor] DEBUG nextflow.file.FileHelper - Path matcher not defined by 'S3FileSystem' file system -- using default default strategy
```

**Comment by pditommaso** [comment]
There are still a lot of way to fetch runtime information. You can use the command `nextflow log <run name> -f <fields>` to get task information of a previous execution, the [trace report](https://www.nextflow.io/docs/latest/tracing.html#trace-report) and the [timeline report](https://www.nextflow.io/docs/latest/tracing.html#timeline-report). 

> However I see a lot of time spent in [Task Monitor].

That's a thread just waiting for tasks completion, I don't think it could be the a bottleneck.

**Comment by tbugfinder** [comment]
As AWS Batch does not list any jobs for hours, the expectation is that NF would also finish very close to that state.

**Comment by pditommaso** [comment]
> As AWS Batch does not list any jobs for hours 

How do you infer that? from the above log seems job completion are recognised in a timeline manner.

**Comment by pditommaso** [comment]
This looks solved.

**Comment by tbugfinder** [comment]
>> As AWS Batch does not list any jobs for hours

>How do you infer that? from the above log seems job completion are recognised in a timeline manner.

Unfortunately this has taken some time to diag on my side.

See attached screenshot of the S3 workdir bucket contents.
![image](https://user-images.githubusercontent.com/15378237/42736068-b09dcf4c-885f-11e8-8bea-b73b8f5dffc4.png)


This is the matching `nextflow log` output.
```
$ time ~/nextflow/nextflow log loving_avogadro -f name,submit,start,complete,duration,realtime,task_id,workdir -F "name =~ '/.*pattern.*/ '"
(pattern)        2018-07-10 01:05:19.143 -       2018-07-10 06:26:22.165 5h 21m 3s       31m 24s 121090  s3://test/work/work/e4/cedf1aa5e367f871b12572b0f4be4e

real    0m38.435s
user    0m38.599s
sys     0m1.378s

```

Comparing the S3 screenshot and nextflow output there is a major delay between job completed and nextflow updating that status.

Should this be a dedicated issue?
Btw, where do I find  the AWS Batch JobId within `nextflow log` ?

**Comment by pditommaso** [comment]
Yes, this looks a different problem. Please open a separate issue including this snapshot. 

> Btw, where do I find the AWS Batch JobId within nextflow log? 

Is reported as `native_id`

**Comment by tbugfinder** [comment]
filed #799
